oauthenticator/globus.py

```diff
@@ -0,0 +1,223 @@
+
+
+try:
+    import globus_sdk
+except:
+    raise ImportError('globus_sdk is not installed, please see '
+                      '"globus-requirements.txt" for using Globus oauth.')
+
+
+class GlobusMixin(OAuth2Mixin):
+    """
+    Use the globus_sdk to get the auth URL.
+    """
+    @return_future
+    def authorize_redirect(self, client=None, callback=None):
+        self.redirect(client.oauth2_get_authorize_url())
+        callback()
+
+
+class GlobusOAuthCallbackHandler(OAuthCallbackHandler):
+    """This extra callback handler is needed to store transfer tokens
+    in auth_state. This is needed since there isn't a sensible way to
+    access the database through `authenticate`, although this may change
+    with the following issue:
+    https://github.com/jupyterhub/jupyterhub/issues/1063
+    globus tokens are saved in the database to persist server
+    restarts and closed browser windows. This ensures consistency -- whenever
+    the user is logged in, they will be able to spawn a notebook with tokens.
+    """
+    @gen.coroutine
+    def get(self):
+        username = yield self.authenticator.get_authenticated_user(self, None)
+
+        if username:
+            user = self.user_from_username(username)
+            self.set_globus_data(user)
+            self.set_login_cookie(user)
+            self.redirect(url_path_join(self.hub.server.base_url, 'home'))
+        else:
+            raise web.HTTPError(403)
+
+    def set_globus_data(self, user):
+        user.auth_state = {
+            'globus_data': self.authenticator.globus_data
+        }
+        self.db.commit()
+
+
+class GlobusLoginHandler(OAuthLoginHandler, GlobusMixin):
+    """
+    The login handler sets the scope and provides the redirect URL.
+    The scope can be modified from the config.
+    """
+
+    def get(self):
+        redirect_uri = self.authenticator.get_callback_url(self)
+        client = self.authenticator.globus_portal_client()
+        client.oauth2_start_flow(
+            redirect_uri,
+            requested_scopes=' '.join(self.authenticator.scope),
+            refresh_tokens=self.authenticator.allow_refresh_tokens
+        )
+        self.log.info('globus redirect: %r', redirect_uri)
+        self.authorize_redirect(client)
+
+
+class GlobusOAuthenticator(OAuthenticator):
+    """The Globus OAuthenticator handles both authorization and passing
+    transfer tokens to the spawner. """
+
+    login_service = 'Globus'
+    login_handler = GlobusLoginHandler
+    callback_handler = GlobusOAuthCallbackHandler
+
+    identity_provider = Unicode(help="""Restrict which institution a user
+    can use to login (GlobusID, University of Hogwarts, etc.). This should
+    be set in the app at developers.globus.org, but this acts as an additional
+    check to prevent unnecessary account creation.""").tag(config=True)
+
+    def _identity_provider_default(self):
+        return os.getenv('IDENTITY_PROVIDER', 'globusid.org')
```

Comment:
If an admin wanted to change which Identity is used by their users logging in, they can configure it in their [app](developers.globus.org). However, this serves as a nice additional check that can easily be seen in the `jupyterhub_config.py` file.